ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala

```diff
@@ -0,0 +1,66 @@
+package com.daml.ledger.participant.state.kvutils.committer
+
+import com.codahale.metrics
+import com.codahale.metrics.Timer
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlLogEntry,
+  DamlLogEntryId,
+  DamlStateKey,
+  DamlStateValue
+}
+import com.daml.ledger.participant.state.v1.ParticipantId
+import com.digitalasset.daml.lf.data.Time
+import org.slf4j.{Logger, LoggerFactory}
+
+trait Committer[Submission, PartialResult] {
+
+  /** A kvutils committer is composed of individual commit steps which:
+    *  - carry a partial result from one step to another
+    *  - can access daml state via [[Context.get]]
+    *  - can update daml state via [[Context.set]]
+    *  - can end the commit via [[Context.done]], which skips the rest of the remaining steps
+    */
+  type Step = (Context, PartialResult) => PartialResult
+  def steps: Iterable[Step]
+
+  /** The initial partial result passed to first step. */
+  def init(subm: Submission): PartialResult
+
+  val logger: Logger = LoggerFactory.getLogger(this.getClass)
+  val metricsRegistry: metrics.MetricRegistry =
+    metrics.SharedMetricRegistries.getOrCreate("kvutils")
+  def metricsName(metric: String): String = metrics.MetricRegistry.name(this.getClass, metric)
+  private val runTimer: Timer = metricsRegistry.timer(metricsName("run-timer"))
+
+  /** A committer can `run` a submission and produce a log entry and output states. */
+  def run(
+      entryId: DamlLogEntryId,
+      recordTime: Time.Timestamp,
+      submission: Submission,
+      participantId: ParticipantId,
+      inputState: DamlStateMap): (DamlLogEntry, Iterable[(DamlStateKey, DamlStateValue)]) =
+    runTimer.time { () =>
+      var result: Option[DamlLogEntry] = None
+      val ctx = new Context {
+        override def getRecordTime: Time.Timestamp = recordTime
+        override def getParticipantId: ParticipantId = participantId
+        override def inputs: DamlStateMap = inputState
+        override def done[Void](logEntry: DamlLogEntry): Void = {
+          result = Some(logEntry)
+          // Return "null", giving us the assertion and a compiler warning if the committer
```

Comment:
`Void`, `Nothing`'s slightly weaker sibling ;)

Reply:
btw, any idea why this construct makes the scala compiler warn of dead code following the `done()` call? I cannot figure out why, but I like it.

Also better less hacky approach to this would be welcome :).

Reply:
Wait, I completely missed that this `Void` isn't actually `java.lang.Void`. So basically it's the `Nothing` check. In fact, if you look at what type is inferred for `PackageCommitter#finish`, you will find that it is indeed `Nothing`.

Reply:
In fact, this should result in a null pointer exception, since the use site in `PackageCommitter` doesn't specify a type parameter.

```
scala> def n: Nothing = null.asInstanceOf[Nothing]
n: Nothing

scala> def mayFail[F] = null.asInstanceOf[F]
mayFail: [F]=> F

scala> n
java.lang.NullPointerException
  at .$print$lzycompute(<console>:9)

scala> mayFail[Int]
res4: Int = 0

scala> mayFail
java.lang.NullPointerException
  at .$print$lzycompute(<console>:9)

scala> mayFail[String]
res6: String = null
```

Reply:
Dropped all this sillyness and used an ADT (Continue/Stop).